ui/src/notebooks/components/header/Submit.tsx

```diff
@@ -79,9 +79,37 @@ export const Submit: FC = () => {
               requirements,
             })
           } else if (pipe.type === 'data') {
-            const {bucketName} = pipe
+            const {bucketName, field, measurement, tags} = pipe
 
-            const text = `from(bucket: "${bucketName}")|>range(start: v.timeRangeStart, stop: v.timeRangeStop)`
+            let text = `from(bucket: "${bucketName}")|>range(start: v.timeRangeStart, stop: v.timeRangeStop)`
+            if (measurement) {
+              text += `|> filter(fn: (r) => r["_measurement"] == "${measurement}")`
+            }
+            if (field) {
+              text += `|> filter(fn: (r) => r["_field"] == "${field}")`
+            }
+            if (tags && Object.keys(tags)?.length > 0) {
+              Object.keys(tags)
+                .filter(tagName => !!tags[tagName])
+                .forEach(tagName => {
+                  const tagValues = tags[tagName]
+                  const values = tagValues as any[]
+                  if (values.length === 1) {
+                    text += `|> filter(fn: (r) => r["${tagName}"] == "${values[0]}")`
+                  } else {
+                    values.forEach((val, i) => {
+                      if (i === 0) {
+                        text += `|> filter(fn: (r) => r["${tagName}"] == "${val}"`
+                      }
+                      if (values.length - 1 === i) {
+                        text += ` or r["${tagName}"] == "${val}")`
+                      } else {
+                        text += ` or r["${tagName}"] == "${val}"`
+                      }
+                    })
+                  }
+                })
+            }
```

Comment:
It feels like this can be abstracted into a utility function?

Reply:
maybe we could add it to the pipe registration too, so that they can all define how they wanna make their own queries. i think previous_results is preventing this currently, but worth a look